File: contest_6/E.cpp
#include <iostream>
#include <deque>
#include <cmath>
#include <limits>
#include <vector>
#pragma GCC optimize("Ofast")

int min_digit_sum_divisible_by_k(int k) {
    if (k == 1) {
        return 1;
    }
    int inf = std::numeric_limits<int>::max();
    std::vector<int> dp(k, inf);
    std::deque<int> d;
    for (int i = 1; i < 10; i++) {
        dp[i % k] = std::min(dp[i % k], i);
        d.push_back(i % k);
    }
    while (!d.empty()) {
        int p = d.front();
        d.pop_front();
        for (int digit = 0; digit < 10; digit++) {
            int new_p = (p * 10 + digit) % k;
            int new_sum = dp[p] + digit;
            if (new_sum < dp[new_p]) {
                dp[new_p] = new_sum;
                d.push_back(new_p);
            }
        }
    }
    if (dp[0] != inf) {
        return dp[0];
    } else {
        return k;
    }
}

int main() {
    int k;
    std::cin >> k;
    std::cout << min_digit_sum_divisible_by_k(k) << std::endl;
    return 0;
}

File: contest_6/E.py
from collections import deque
from math import inf


def min_digit_sum_divisible_by_k(k):
    if k == 1:
        return 1
    dp = [inf] * k
    d = deque()
    # определение цифр, которые можем использовать (все цифры меньше k)
    for i in range(1, 10):
        dp[i % k] = min(dp[i % k], i)
        d.append(i % k)
    while d:
        p = d.popleft()  # Извлекаем текущую последовательность найденных цифр из очереди
        for digit in range(10):  # Перебираем все возможные цифры
            new_p = (p * 10 + digit) % k  # Вычисляем новое значение остатка от деления
            new_sum = dp[p] + digit  # Вычисляем новую сумму цифр
            if new_sum < dp[new_p]:  # Если новая сумма меньше текущего значения в dp для этого остатка
                dp[new_p] = new_sum  # Обновляем значение в dp
                d.append(new_p)  # Добавляем новый остаток в очередь d
    if dp[0] != inf:
        return dp[0]
    else:
        return k


# Запускаем функцию с вводом k и выводим результат
print(min_digit_sum_divisible_by_k(int(input())))
